src/lsp-server/src/backend.rs

```diff
@@ -297,6 +416,62 @@ impl LanguageServer for Backend {
         Ok(None)
     }
 
+    /// Completion implementation.
+    async fn completion(&self, params: CompletionParams) -> Result<Option<CompletionResponse>> {
+        let _uri = params.text_document_position.text_document.uri;
+        let _position = params.text_document_position.position;
+
+        let content = self.content.lock().await;
+        let content = content.get(&_uri);
+
+        if let Some(content) = content {
+            // Get the lex token.
+            let lex_results = lexer::lex(&content.to_string())
+                .map_err(|_| build_error("Error getting lex tokens."))?;
+            let offset = position_to_offset(_position, content)
+                .ok_or(build_error("Error getting completion offset."))?;
+            let mut last_keyword: Option<Keyword> = None;
+            lex_results.iter().find(|x| {
+                if x.offset < offset {
+                    match x.kind {
+                        Token::Keyword(k) => match k {
+                            Keyword::Select => {
+                                last_keyword = Some(k);
```

Comment:
You might need `.filter_map(..).last()` here instead. Using `find` like and mutating external state is not common practice.